torchdata/stateful_dataloader/ibm_rescalable.py

```diff
@@ -0,0 +1,614 @@
+            for i in range(self.n_logicals):
+                self.data.append(deepcopy(self.dataset))
+                self.data[-1].worldsize = n_logical_shards
+                self.data[-1].rank = self.logicals_owned[i]
+                self.data[-1].local_worldsize = 1
+                self.data[-1].datapath = self.datapath
+                self.data[-1].verbose = self.rank == 0
+                if self.verbose:
+                    logging.info(
+                        f"Worker {self.rank} assembled logical shard {self.logicals_owned[i]}, {i+1} of {self.n_logicals}"
+                    )
+            [d.setup() for d in self.data]
+
+    def __iter__(self):
+        self.setup()
+        # Grab one item at a time, iterating over owned logical shards
+        data = [iter(d) for d in self.data]
+        while True:
+            ind = self.current_reader
+            # Read doc
+            out = next(data[ind])
+            # Update state
+            self.current_reader = (self.current_reader + 1) % self.n_logicals
+            yield out
+
+    def state_dict(self):
+        self.setup()
+        # Recursive fetch
+        logical_shard_states = [d.state_dict() for d in self.data]
+        assert len(logical_shard_states) > 0, f"Worker {self.rank} owns no shards???"
+        # Flip list[dict[Any]] to dict[list[Any]]
+        state_dict = {k: [d[k] for d in logical_shard_states] for k in logical_shard_states[0].keys()}
+        state_dict.update(_StatefulDataset.state_dict(self))
+
+        # Convert to tensor form
+        out = {}
+        for k, v in state_dict.items():
+            v = torch.tensor(v)
+            if len(v.shape) == 0:
+                k = k + ".scalar"
+                v = v.unsqueeze(0)
+            out[k] = v
+
+        return out
+
+    def load_state_dict(self, state_dict):
+        self.setup()
+
+        # Convert back to lists and scalars
+        def detorchify(k, v):
+            v = v.tolist()
+            if ".scalar" in k:
+                k = k[:-7]
+                v = v[0]
+            return k, v
+
+        plain_dict = {}
+        for k, v in state_dict.items():
+            k, v = detorchify(k, v)
+            plain_dict[k] = v
+        state_dict = plain_dict
+
+        # Assemble logical shard states
+        # TODO: how is this handling non-resharding state_params when resharding???
+        _StatefulDataset.load_state_dict(self, state_dict)
+        # Remove all non-resharding state
+        [state_dict.pop(self.statename(n)) for n in self.state_params]
+        # Flip dict[list[any]] to list[dict[any]]
+        logical_shard_states = [{k: v[i] for k, v in state_dict.items()} for i in range(self.n_logicals)]
+
+        # Load values
+        for i in range(self.n_logicals):
+            self.data[i].load_state_dict(logical_shard_states[i])
+
+
+#### -------------------------    CHECKPOINT FUNCTIONS    ------------------------- ####
+
+
+def __pop_dstate(state, device_mesh, placements):
```

Comment:
We should create standard utilities to get these in torchdata  https://github.com/pytorch/data/issues/1337